lib/ansible/modules/web_infrastructure/jenkins_node.py

```diff
@@ -0,0 +1,370 @@
+#!/usr/bin/python
+#
+# Copyright: (c) Ansible Project
+#
+# Simplified BSD License (see licenses/simplified_bsd.txt or https://opensource.org/licenses/BSD-2-Clause)
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: jenkins_node
+short_description: Jenkins Node/Executor Operation.
+description:
+ - Uses Jenkins REST API to manage jenkins node/Executor operations.
+ - Change node label via append/replace.
+ - Bring node online or offline with message.
+ - Wait for node to become idle and then proceed with the ansible tasks.
+
+version_added: "2.10"
+requirements:
+    - "python-jenkins >= 1.4.0"
+options:
+    host:
+        description:
+            - Jenkins node/Executor name.
+        type: str
+        required: true
+    jenkins_server:
+        description:
+            - Jenkins server url(incl http/https, port).
+        type: str
+        required: true
+    jenkins_username:
+        description:
+            - Authenticated username with Jenkins server.
+        type: str
+        required: true
+    jenkins_token:
+        description:
+            - Authenticated token with Jenkins server.
+        type: str
+        required: true
+    wait_until_free:
+        description:
+            - Check and halt play execution if node/Executor is running jenkins job.
+            - halt makes present task sleep for 60 seconds.
+        type: bool
+        required: false
+    online:
+        description:
+            - state tells whether to bring node offline or online
+            - msg message on Jenkins UI when node is bought offline.
+        type: dict
+        required: false
+    label:
+        description:
+            - Modify jenkins node label name.
+            - Provides ability to add new label, add suffix or prefix label, remove label,
+              remove suffix or prefix label.
+        type: dict
+        required: false
+    delete:
+        description:
+            - Delete  node from jenkins master.
+        type: bool
+        required: false
+seealso:
+    - module: jenkins_job
+    - module: jenkins_plugin
+author: "Nalin Garg (@nalingarg2)"
+'''
+
+EXAMPLES = '''
+
+-   name: Add new label.
+    jenkins_node:
+        host: myhost
+        jenkins_server: https://jenkins.example.com
+        jenkins_username: test_username
+        jenkins_token: test_token
+        label:
```

Comment:
@stpierre I can understand why it seems complicated for the simple act of giving nodes "label" on Jenkins. The reason for complications lies in Jenkins which acts as "scheduler" and "load balancer". This module is aimed to make actions easier for both jenkins and ansible admins.
In an active Jenkins cluster (Jenkins scheduling builds), we would want:
1. To change existing labels so that Jenkins doesn't schedule new builds on existing labels. This is the only way Ansible and Jenkins can co-exist. 
2. Each label change is marked as "config change". From Jenkins admin's perspective, when the diff is seen on a given node, we want to minimize the ansible change(helps in debugging Jenkins).
PS: One jenkins node might have many labels.